### PR TITLE
Port jormungandr style

### PR DESCRIPTION
**Pull request by paweljakubas**
Task:
#24 

Done:
1. Ported all needed interfaces from cardano-wallet
2. Added deriveStakingPrivatekey to DelegationAddress interface
3. Added unit tests

### CODE REVIEW

File: src/Cardano/Address.hs [hunk 1]
```diff
@@ -130,6 +134,19 @@ class PaymentAddress key
...
+    --
+    -- Deriving staking keys for something else than the initial account is not
+    -- recommended and can lead to incompatibility with existing wallet softwares
+    -- (Daedalus, Yoroi, Adalite...).
+    --
+    -- @since 2.0.0
+    deriveStakingPrivateKey
+        :: key 'AccountK XPrv
+        -> key 'StakingK XPrv
+
```
Comment (KtorZ) [inline src/Cardano/Address.hs:149]: I don't think this is the right place to have this one. I'd suggest creating another type-class to capture this constraint as

```hs
class StakingDerivation key where
   ...
```

Very much like we have `SoftDerivation` and `HardDerivation`.

File: src/Cardano/Address/Style/Jormungandr.hs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+-- https://github.com/input-output-hk/implementation-decisions/blob/e2d1bed5e617f0907bc5e12cf1c3f3302a4a7c42/text/1852-hd-chimeric.md
+--
+-- It indicates that the subtree of this node is used according to this
+-- specification.
+--
+-- Hardened derivation is used at this level.
+purposeIndex :: Word32
+purposeIndex = 0x8000073c
+
+-- | One master node (seed) can be used for unlimited number of independent
```
Comment (KtorZ) [inline src/Cardano/Address/Style/Jormungandr.hs:296]: Note that we don't want all these constants to show up in the documentation. Haddock will automatically prune these **if and only if** they have no docstring attached. So here, we need to replace `-- | ...` with `-- ...` for everything we don't want to be exposed.

File: src/Cardano/Address/Style/Jormungandr.hs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+instance DelegationAddress Jormungandr where
+    delegationAddress discrimination paymentKey stakingKey = unsafeMkAddress $
+        invariantSize expectedLength $ BL.toStrict $ runPut $ do
+            putWord8 firstByte
+            putByteString . getPublicKey . getKey $ paymentKey
+            putByteString . getPublicKey . getKey $ stakingKey
+      where
+          firstByte = case addressDiscrimination @Jormungandr discrimination of
+              RequiresNetworkTag -> 0x84
+              RequiresNoTag -> 0x04
```
Comment (KtorZ) [inline src/Cardano/Address/Style/Jormungandr.hs:250]: These constants should be the value behind `networkTag @Jormungandr`. So here, I'd suggest to simply export a single smart constructor as:

```hs
incentivizedTestnet :: NetworkDiscriminat Jormungandr
incentivizedTestnet = (RequiresNetworkTag, 0x84)
``` 

And, simply use `networkTag @Jormungandr` for the first byte. The purpose of the `HasNetworkDiscriminant` abstraction is to capture these constants and offer a consistant API to the caller.

File: src/Cardano/Address/Style/Jormungandr.hs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+jormungandrTestnet :: NetworkDiscriminant Jormungandr
+jormungandrTestnet = (RequiresNetworkTag, NetworkTag 0)
+
+{-------------------------------------------------------------------------------
+                                 Key generation
+-------------------------------------------------------------------------------}
+
+-- | Size, in bytes, of a public key (without chain code)
+publicKeySize :: Int
+publicKeySize = 32
```
Comment (KtorZ) [inline src/Cardano/Address/Style/Jormungandr.hs:274]: Same as for Shelley, it'd be better to pull the constant from cryptonite here. https://hackage.haskell.org/package/cryptonite-0.26/docs/Crypto-PubKey-Ed25519.html#v:publicKeySize

File: src/Cardano/Address.hs (hunk 1 above)
Comment (paweljakubas) [reply src/Cardano/Address.hs]: done

File: src/Cardano/Address/Style/Jormungandr.hs (hunk 2 above)
Comment (paweljakubas) [reply src/Cardano/Address/Style/Jormungandr.hs]: done

File: src/Cardano/Address/Style/Jormungandr.hs (hunk 3 above)
Comment (paweljakubas) [reply src/Cardano/Address/Style/Jormungandr.hs]: done

File: src/Cardano/Address/Style/Jormungandr.hs (hunk 4 above)
Comment (paweljakubas) [reply src/Cardano/Address/Style/Jormungandr.hs]: done

File: src/Cardano/Address/Style/Jormungandr.hs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+            putWord8 (invariantNetworkTag 255 firstByte)
+            putByteString (getPublicKey $ getKey k)
+      where
+          firstByte = networkTag @Jormungandr discrimination
+          expectedLength = 1 + publicKeySize
+
+instance DelegationAddress Jormungandr where
+    delegationAddress discrimination paymentKey stakingKey = unsafeMkAddress $
+        invariantSize expectedLength $ BL.toStrict $ runPut $ do
+            putWord8 (invariantNetworkTag 255 $ NetworkTag $ firstByte + 1)
```
Comment (KtorZ) [inline src/Cardano/Address/Style/Jormungandr.hs:260]: `+1` :thinking:  ?

Comment (paweljakubas) [reply src/Cardano/Address/Style/Jormungandr.hs:260]: for paymentAddress singleAddress is used, for delegationAddress - groupedAddress : https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs#L343-L344 - if I understand well

Comment (KtorZ) [reply src/Cardano/Address/Style/Jormungandr.hs:260]: :+1:

**Pull request merged**